File: Emulator/src/main/core/game/interaction/InteractPlugin.java
package core.game.interaction;

import core.game.container.Container;
import core.game.event.InteractionEvent;
import core.game.node.Node;
import core.game.node.entity.impl.PulseType;
import core.game.node.entity.npc.NPC;
import core.game.node.entity.player.Player;
import core.game.node.item.Item;
import core.game.node.scenery.Scenery;
import core.game.system.task.Pulse;
import core.game.world.GameWorld;
import core.net.packet.PacketRepository;
import core.net.packet.context.InteractionOptionContext;
import core.net.packet.out.InteractionOption;
import core.tools.Log;

import static core.api.ContentAPIKt.log;

/**
 * The type Interact plugin.
 */
public class InteractPlugin {

    private Option[] options = new Option[8];

    private final Node node;

    private boolean initialized;

    /**
     * Instantiates a new Interact plugin.
     *
     * @param node the node
     */
    public InteractPlugin(Node node) {
        this.node = node;
    }

    /**
     * Send option.
     *
     * @param node  the node
     * @param index the index
     * @param name  the name
     */
    public static void sendOption(Node node, int index, String name) {
        if (!(node instanceof Player)) {
            return;
        }
        PacketRepository.send(InteractionOption.class, new InteractionOptionContext((Player) node, index, name, false));
    }

    /**
     * Handle.
     *
     * @param player the player
     * @param option the option
     */
    public void handle(final Player player, final Option option) {
        try {
            if (player.getLocks().isInteractionLocked() || option == null) {
                return;
            }
            player.debug("Received interaction request " + option.getName());
            boolean hasHandler = option.getHandler() != null;
            boolean walk = hasHandler && option.getHandler().isWalk();
            if (!walk && hasHandler && option.getHandler().isWalk(player, node)) {
                walk = true;
            }
            if (!hasHandler || walk) {
                handleWalkOption(player, option, PulseType.STANDARD);
            } else if (hasHandler) {
                player.debug("Option handler being used=" + option.getHandler().getClass().getSimpleName());
                handleDefaultOption(player, option, PulseType.STANDARD);
            } else {
                player.getPulseManager().runUnhandledAction(player, PulseType.STANDARD);
            }
            player.dispatch(new InteractionEvent(node, option.getName().toLowerCase()));
        } catch (Exception e) {
            e.printStackTrace();
            log(this.getClass(), Log.ERR, this.getClass().getName() + e.getMessage());
        }
    }

    /**
     * Handle item option.
     *
     * @param player    the player
     * @param option    the option
     * @param container the container
     */
    public void handleItemOption(final Player player, final Option option, final Container container) {
        if (player.getLocks().isInteractionLocked()) {
            return;
        }
        player.getPulseManager().clear(PulseType.STANDARD);
        GameWorld.getPulser().submit(new Pulse(1, player) {
            @Override
            public boolean pulse() {
                try {
                    if (player.getLocks().isInteractionLocked() || player.getZoneMonitor().interact(node, option)) {
                        return true;
                    }
                    if (InteractionListeners.run(node.getId(), IntType.ITEM, option.getName(), player, node)) {
                        return true;
                    }
                    if (option.getHandler() == null || !option.getHandler().handle(player, node, option.getName().toLowerCase())) {
                        player.getPacketDispatch().sendMessage("Nothing interesting happens.");
                    }
                    if (option.getHandler() != null) {
                        player.debug("Using item handler " + option.getHandler().getClass().getSimpleName());
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    log(this.getClass(), Log.ERR, this.getClass().getName() + e.getMessage());
                }
                return true;
            }
        });
    }

    /**
     * Handle invalid interaction.
     *
     * @param player the player
     * @param node   the node
     * @param option the option
     */
    public static void handleInvalidInteraction(final Player player, final Node node, final Option option) {
        if (node == null) {
            return;
        }
        if (node.getLocation() != null) {
            if (player.getLocks().isMovementLocked()) {
                return;
            }
            player.getPulseManager().run(new MovementPulse(player, node) {
                @Override
                public boolean pulse() {
                    try {
                        player.faceLocation(node.getFaceLocation(player.getLocation()));
                        if (player.getLocks().isInteractionLocked() || player.getZoneMonitor().interact(node, option)) {
                            return true;
                        }
                        player.getPacketDispatch().sendMessage("Nothing interesting happens.");
                    } catch (Exception e) {
                        e.printStackTrace();
                        log(this.getClass(), Log.ERR, this.getClass().getName() + e.getMessage());
                    }
                    return true;
                }
            }, PulseType.STANDARD);
        } else {
            player.getPulseManager().runUnhandledAction(player, PulseType.STANDARD);
        }
    }

    private void handleWalkOption(final Player player, final Option option, PulseType pulseType) {
        if (node.getLocation() == null) {
            player.getPulseManager().runUnhandledAction(player, pulseType);
            return;
        }
        if (player.getLocks().isMovementLocked()) {
            player.getPulseManager().clear(pulseType);
            return;
        }
        player.getPulseManager().run(new MovementPulse(player, node, option.getHandler()) {
            @Override
            public boolean pulse() {
                try {
                    player.faceLocation(node.getFaceLocation(player.getLocation()));
                    if (player.getLocks().isInteractionLocked() || player.getZoneMonitor().interact(node, option)) {
                        return true;
                    }
                    if (option == null || option.getHandler() == null || !option.getHandler().handle(player, node, option.getName().toLowerCase())) {
                        player.getPacketDispatch().sendMessage("Nothing interesting happens.");
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    log(this.getClass(), Log.ERR, this.getClass().getName() + e.getMessage());
                }
                return true;
            }
        }, pulseType);
    }

    private void handleDefaultOption(final Player player, final Option option, PulseType pulseType) {
        if (!option.getHandler().isDelayed(player)) {
            if (player.getZoneMonitor().interact(node, option)) {
                return;
            }
            player.getProperties().getCombatPulse().stop();
            if (!option.getHandler().handle(player, node, option.getName().toLowerCase())) {
                player.getPulseManager().runUnhandledAction(player, pulseType);
            }
            return;
        }
        player.getPulseManager().run(new Pulse(1, player, node) {
            @Override
            public boolean pulse() {
                if (player.getLocks().isInteractionLocked() || player.getZoneMonitor().interact(node, option)) {
                    return true;
                }
                if (!option.getHandler().handle(player, node, option.getName().toLowerCase())) {
                    player.getPacketDispatch().sendMessage("Nothing interesting happens.");
                }
                return true;
            }
        }, pulseType);
    }

    /**
     * Init.
     *
     * @param nodeId the node id
     * @param names  the names
     */
    public void init(int nodeId, String... names) {
        options = new Option[names.length];
        for (int i = 0; i < options.length; i++) {
            String name = names[i];
            if (name != null && !name.equals("null")) {
                set(new Option(name, i).setHandler(Option.defaultHandler(node, nodeId, name)));
            }
            continue;
        }
    }

    /**
     * Sets default.
     */
    public void setDefault() {
        if (initialized) {
            return;
        }
        if (node instanceof Player) {
            for (int i = 0; i < options.length; i++) {
                remove(i);
            }
            set(Option._P_FOLLOW);
            set(Option._P_TRADE);
            set(Option._P_ASSIST);
        } else if (node instanceof NPC) {
            NPC npc = (NPC) node;
            init(npc.getId(), npc.getDefinition().getOptions());
        } else if (node instanceof Scenery) {
            Scenery scenery = (Scenery) node;
            init(scenery.getId(), scenery.getDefinition().getOptions());
        } else if (node instanceof Item) {
            Item item = (Item) node;
            if (item.getLocation() != null) {
                init(item.getId(), item.getDefinition().groundOptions);
            } else {
                init(item.getId(), item.getDefinition().getInventoryOptions());
            }
        } else {
            throw new IllegalStateException("Unsupported node type - " + node);
        }
        initialized = true;
    }

    /**
     * Set.
     *
     * @param option the option
     */
    public void set(Option option) {
        options[option.getIndex()] = option;
        sendOption(node, option.getIndex(), option.getName());
    }

    /**
     * Remove boolean.
     *
     * @param option the option
     * @return the boolean
     */
    public boolean remove(Option option) {
        if (options[option.getIndex()] == option) {
            remove(option.getIndex());
            return true;
        }
        return false;
    }

    /**
     * Remove.
     *
     * @param index the index
     */
    public void remove(int index) {
        if (options[index] == null) {
            return;
        }
        options[index] = null;
        sendOption(node, index, "null");
    }

    /**
     * Get option.
     *
     * @param index the index
     * @return the option
     */
    public Option get(int index) {
        return options[index];
    }

    /**
     * Get options option [ ].
     *
     * @return the option [ ]
     */
    public Option[] getOptions() {
        return options;
    }

    /**
     * Is initialized boolean.
     *
     * @return the boolean
     */
    public boolean isInitialized() {
        return initialized;
    }

    /**
     * Sets initialized.
     *
     * @param initialized the initialized
     */
    public void setInitialized(boolean initialized) {
        this.initialized = initialized;
    }

}

File: Emulator/src/main/core/game/interaction/InteractPlugin.kt
package core.game.interaction

import core.api.log
import core.game.container.Container
import core.game.event.InteractionEvent
import core.game.interaction.InteractionListeners.run
import core.game.node.Node
import core.game.node.entity.impl.PulseType
import core.game.node.entity.npc.NPC
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.game.node.scenery.Scenery
import core.game.system.task.Pulse
import core.game.world.GameWorld.Pulser
import core.net.packet.OutgoingContext.InteractionOptionContext
import core.net.packet.PacketRepository.send
import core.net.packet.out.InteractionOption
import core.tools.Log
import java.util.*

/**
 * Handles interaction logic for a node.
 */
class InteractPlugin(private val node: Node) {

    /**
     * The list of available options for this node.
     */
    var options: Array<Option?> = arrayOfNulls(8)
        private set

    var isInitialized: Boolean = false

    /**
     * Handles a general interaction option click.
     *
     * @param player the player interacting
     * @param option the selected interaction option
     */
    fun handle(player: Player, option: Option?) {
        try {
            if (player.locks.isInteractionLocked() || option == null) return

            player.debug("Received interaction request: ${option.name}")
            val handler = option.handler
            val shouldWalk = handler?.isWalk == true || handler?.isWalk(player, node) == true

            when {
                handler == null || shouldWalk -> handleWalkOption(player, option, PulseType.STANDARD)
                else -> {
                    player.debug("Option handler used: ${handler.javaClass.simpleName}")
                    handleDefaultOption(player, option, PulseType.STANDARD)
                }
            }

            player.dispatch(InteractionEvent(node, option.name.lowercase(Locale.getDefault())))

        } catch (e: Exception) {
            e.printStackTrace()
            log(javaClass, Log.ERR, "${javaClass.name}: ${e.message}")
        }
    }

    /**
     * Handles using an item on this node.
     *
     * @param player the player using the item
     * @param option the selected item option
     * @param container the item container (optional)
     */
    fun handleItemOption(player: Player, option: Option, container: Container?) {
        if (player.locks.isInteractionLocked()) return

        player.pulseManager.clear(PulseType.STANDARD)

        Pulser.submit(object : Pulse(1, player) {
            override fun pulse(): Boolean {
                try {
                    if (player.locks.isInteractionLocked() || player.zoneMonitor.interact(node, option)) return true

                    if (run(node.id, IntType.ITEM, option.name, player, node)) return true

                    val handler = option.handler
                    if (handler == null || !handler.handle(player, node, option.name.lowercase(Locale.getDefault()))) {
                        player.packetDispatch.sendMessage("Nothing interesting happens.")
                    } else {
                        player.debug("Using item handler: ${handler.javaClass.simpleName}")
                    }

                } catch (e: Exception) {
                    e.printStackTrace()
                    log(javaClass, Log.ERR, "${javaClass.name}: ${e.message}")
                }
                return true
            }
        })
    }

    /**
     * Handles walking toward the node and executing the option.
     *
     * @param player the player moving toward the node
     * @param option the interaction option
     * @param pulseType the pulse type to use
     */
    private fun handleWalkOption(player: Player, option: Option, pulseType: PulseType) {
        if (node.location == null) {
            player.pulseManager.runUnhandledAction(player, pulseType)
            return
        }
        if (player.locks.isMovementLocked()) {
            player.pulseManager.clear(pulseType)
            return
        }

        player.pulseManager.run(object : MovementPulse(player, node, option.handler) {
            override fun pulse(): Boolean {
                try {
                    player.faceLocation(node.getFaceLocation(player.location))
                    if (player.locks.isInteractionLocked() || player.zoneMonitor.interact(node, option)) return true

                    val handler = option.handler
                    if (handler == null || !handler.handle(player, node, option.name.lowercase(Locale.getDefault()))) {
                        player.packetDispatch.sendMessage("Nothing interesting happens.")
                    }

                } catch (e: Exception) {
                    e.printStackTrace()
                    log(javaClass, Log.ERR, "${javaClass.name}: ${e.message}")
                }
                return true
            }
        }, pulseType)
    }

    /**
     * Handles immediate (non-walk) interaction with a node.
     *
     * @param player the interacting player
     * @param option the selected option
     * @param pulseType the pulse type to use
     */
    private fun handleDefaultOption(player: Player, option: Option, pulseType: PulseType) {
        val handler = option.handler ?: return

        if (!handler.isDelayed(player)) {
            if (player.zoneMonitor.interact(node, option)) return

            player.properties.combatPulse.stop()
            if (!handler.handle(player, node, option.name.lowercase(Locale.getDefault()))) {
                player.pulseManager.runUnhandledAction(player, pulseType)
            }
        } else {
            player.pulseManager.run(object : Pulse(1, player, node) {
                override fun pulse(): Boolean {
                    if (player.locks.isInteractionLocked() || player.zoneMonitor.interact(node, option)) return true
                    if (!handler.handle(player, node, option.name.lowercase(Locale.getDefault()))) {
                        player.packetDispatch.sendMessage("Nothing interesting happens.")
                    }
                    return true
                }
            }, pulseType)
        }
    }

    /**
     * Initializes the interaction options for a given node id and option names.
     *
     * @param nodeId the ID of the node
     * @param names the names of interaction options
     */
    fun init(nodeId: Int, vararg names: String?) {
        options = arrayOfNulls(names.size)
        for (i in names.indices) {
            val name = names[i]
            if (!name.isNullOrBlank() && name != "null") {
                set(Option(name, i).setHandler(Option.defaultHandler(node, nodeId, name)))
            }
        }
    }

    /**
     * Automatically assigns default options based on node type.
     */
    fun setDefault() {
        if (isInitialized) return

        when (node) {
            is Player -> {
                for (i in options.indices) remove(i)
                set(Option.P_FOLLOW)
                set(Option.P_TRADE)
                set(Option.P_ASSIST)
            }
            is NPC -> init(node.id, *node.definition.options)
            is Scenery -> init(node.id, *node.definition.options)
            is Item -> {
                val definition = node.definition
                val options = if (node.location != null) definition.groundOptions else definition.inventoryOptions
                init(node.id, *options)
            }
            else -> throw IllegalStateException("Unsupported node type: $node")
        }

        isInitialized = true
    }

    /**
     * Sets an interaction option.
     *
     * @param option the option to set
     */
    fun set(option: Option) {
        options[option.index] = option
        sendOption(node, option.index, option.name)
    }

    /**
     * Removes an option instance.
     *
     * @param option the option to remove
     * @return true if it was removed, false otherwise
     */
    fun remove(option: Option): Boolean {
        if (options[option.index] == option) {
            remove(option.index)
            return true
        }
        return false
    }

    /**
     * Removes the option at the specified index.
     *
     * @param index the index to clear
     */
    fun remove(index: Int) {
        if (options[index] == null) return
        options[index] = null
        sendOption(node, index, "null")
    }

    /**
     * Gets the option at the given index.
     *
     * @param index the index to get
     * @return the option or null
     */
    operator fun get(index: Int): Option? = options.getOrNull(index)

    companion object {
        /**
         * Sends an interaction option to client.
         *
         * @param node the node (must be a player)
         * @param index the index of the option
         * @param name the option name
         */
        fun sendOption(node: Node?, index: Int, name: String) {
            if (node is Player) {
                send(InteractionOption::class.java, InteractionOptionContext(node, index, name, false))
            }
        }

        /**
         * Handles fallback behavior for invalid interaction.
         *
         * @param player the player
         * @param node the node being interacted with
         * @param option the selected option
         */
        fun handleInvalidInteraction(player: Player, node: Node?, option: Option?) {
            if (node?.location != null && !player.locks.isMovementLocked()) {
                player.pulseManager.run(object : MovementPulse(player, node) {
                    override fun pulse(): Boolean {
                        try {
                            player.faceLocation(node.getFaceLocation(player.location))
                            if (player.locks.isInteractionLocked() || player.zoneMonitor.interact(node, option?: Option.NULL)) {
                                return true
                            }
                            player.packetDispatch.sendMessage("Nothing interesting happens.")
                        } catch (e: Exception) {
                            e.printStackTrace()
                            log(javaClass, Log.ERR, "${javaClass.name}: ${e.message}")
                        }
                        return true
                    }
                }, PulseType.STANDARD)
            } else {
                player.pulseManager.runUnhandledAction(player, PulseType.STANDARD)
            }
        }
    }
}
